### ISSUE TITLE
Way to mix streaming datasets into LLM Foundry config?

### ISSUE THREAD

**Issue by germanjke** [mosaicml/llm-foundry#332]
In your `streaming` readme you can mix datasets like:

```
# mix C4, github code, and internal datasets
streams = [
  Stream(remote='s3://datasets/c4', proportion=0.4),
  Stream(remote='s3://datasets/github', proportion=0.1),
  Stream(remote='gcs://datasets/my_internal', proportion=0.5),
]

dataset = StreamingDataset(
  streams=streams,
  samples_per_epoch=1e8,
)
```

In your pretrain configs you define datasets like:

```
data_local: ./my-copy-c4
data_remote: # If blank, files must be present in data_local
max_seq_len: 2048
global_seed: 17
```

Can we define many datasets here and mix up them in some ratio?

**Comment by codestar12** [comment]
Hi germanjk,

Yes you can mix streaming datasets. In the yaml for the dataset just define streams label. For example.

```yaml
  train_loader:
    name: text
    drop_last: true
    num_workers: 8
    dataset:
      eos_token_id: ${eos_token_id}
      max_seq_len: ${max_seq_len}
      num_canonical_nodes: ${num_canonical_nodes}
      shuffle: true
      shuffle_algo: py1b
      shuffle_seed: ${global_seed}
      streams:
        c4:
          local: /tmp/dataset/c4/
          proportion: 0.8
          remote: s3:/dataset/c4/
          split: train
        markdown:
          local: /tmp/dataset/markdown/
          proportion: 0.2
          remote: s3://dataset/markdown/
          split: null
```

**Comment by germanjke** [comment]
thanks @codestar12 , want to verify how works proportions? for example my  `/tmp/dataset/markdown` is 10 Gb and 0.2 proportion and my `/tmp/dataset/c4` is 100 Gb and 0.8 proportion. 

How data will be sharded?

**Comment by germanjke** [comment]
i.e. how to merge two full datasets for example?

**Comment by germanjke** [comment]
hi @knighton @codestar12 

can I make construction with 1.0 proportions?
```
  train_loader:
    name: text
    drop_last: true
    num_workers: 8
    dataset:
      eos_token_id: ${eos_token_id}
      max_seq_len: ${max_seq_len}
      num_canonical_nodes: ${num_canonical_nodes}
      shuffle: true
      shuffle_algo: py1b
      shuffle_seed: ${global_seed}
      streams:
        c4:
          local: /tmp/dataset/c4/
          proportion: 1.0
          remote: s3:/dataset/c4/
          split: train
        markdown:
          local: /tmp/dataset/markdown/
          proportion: 1.0
          remote: s3://dataset/markdown/
          split: null
          
```

I will get full c4 and full markdown into one?

**Comment by germanjke** [comment]
```
  train_loader:
    name: text
    drop_last: true
    num_workers: 8
    dataset:
      eos_token_id: ${eos_token_id}
      max_seq_len: ${max_seq_len}
      num_canonical_nodes: ${num_canonical_nodes}
      shuffle: true
      shuffle_algo: py1b
      shuffle_seed: ${global_seed}
      streams:
        c4:
          local: /tmp/dataset/c4/
          repeat: 1.0
          remote: s3:/dataset/c4/
          split: train
        markdown:
          local: /tmp/dataset/markdown/
          repeat: 1.0
          remote: s3://dataset/markdown/
          split: null
```

Or this way is better? I'm right how repeat works? Dataset will be `dataset_size*repeat` samples?

So let's see some example:

My dataset A is 100 samples
My dataset B is 200 samples 

If I put them with both `repeat` 1.0 it will 100 + 200 samples, right?

If i put them with 0.2 repeat for A and 0.1 for B it will be 100 * 0.2 + 200 * 0.1 = 20 samples from A and 20 samples from B = 40 sampels, right?

**Comment by germanjke** [comment]
@codestar12 @knighton

**Comment by germanjke** [comment]
1 more question about proportions:

how it will be works?

```
train_loader:
  name: text
  drop_last: true
  num_workers: 8
  dataset:
    eos_token_id: ${eos_token_id}
    max_seq_len: ${max_seq_len}
    num_canonical_nodes: ${num_canonical_nodes}
    shuffle: true
    shuffle_algo: py1b
    shuffle_seed: ${global_seed}
    streams:
      my_dataset_A_300gb:
        local: /tmp/dataset/c4/
        proportion: 0.5
        remote: s3:/dataset/c4/
        split: train
      my_dataset_B_3gb:
        local: /tmp/dataset/markdown/
        proportion: 0.5
        remote: s3://dataset/markdown/
        split: null
```
for example, I have 2 datasets: `my_dataset_A_300gb` and `my_dataset_B_3gb`, which are 300 and 3 GBs.
I defined both proportion 0.5.

But how it will be works if my dataset A in 100x times bigger than B?

It means dataset B will be repeats many times?
